File: typescript/packages/core/src/core/slack/read.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { SlackAccessor } from '../../accessor/slack.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import type { PathSpec } from '../../types.ts'
import { getHistoryJsonl } from './history.ts'
import { getUserProfile } from './users.ts'
import { stripSlash } from '../../util/slash.ts'

const encoder = new TextEncoder()

function fileNotFound(key: string): Error {
  const e = new Error(`ENOENT: ${key}`) as Error & { code: string }
  e.code = 'ENOENT'
  return e
}

export async function read(
  accessor: SlackAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<Uint8Array> {
  const prefix = path.prefix
  let raw = path.original
  if (prefix !== '' && raw.startsWith(prefix)) {
    raw = raw.slice(prefix.length) || '/'
  }
  const key = stripSlash(raw)
  const parts = key.split('/')
  const part0 = parts[0] ?? ''
  const part1 = parts[1] ?? ''
  const part2 = parts[2] ?? ''
  const part3 = parts[3] ?? ''

  if (parts.length === 4 && (part0 === 'channels' || part0 === 'dms') && part3 === 'chat.jsonl') {
    if (index === undefined) throw fileNotFound(key)
    const parentKey = `${prefix}/${part0}/${part1}`
    const lookup = await index.get(parentKey)
    if (lookup.entry === undefined || lookup.entry === null) {
      throw fileNotFound(key)
    }
    return await getHistoryJsonl(accessor, lookup.entry.id, part2)
  }

  if (parts.length === 5 && (part0 === 'channels' || part0 === 'dms') && part3 === 'files') {
    if (index === undefined) throw fileNotFound(key)
    const virtualKey = `${prefix}/${key}`
    const lookup = await index.get(virtualKey)
    if (lookup.entry === undefined || lookup.entry === null) {
      throw fileNotFound(key)
    }
    const url = lookup.entry.extra.url_private_download
    if (typeof url !== 'string' || url === '') {
      throw fileNotFound(key)
    }
    if (accessor.transport.downloadFile === undefined) {
      throw new Error('slack: transport does not support file download')
    }
    return await accessor.transport.downloadFile(url)
  }

  if (parts.length === 2 && part0 === 'users') {
    if (index === undefined) throw fileNotFound(key)
    const virtualKey = `${prefix}/${key}`
    const lookup = await index.get(virtualKey)
    if (lookup.entry === undefined || lookup.entry === null) {
      throw fileNotFound(key)
    }
    const user = await getUserProfile(accessor, lookup.entry.id)
    return encoder.encode(JSON.stringify(user))
  }

  throw fileNotFound(key)
}

File: python/mirage/core/slack/read.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import json

from mirage.accessor.slack import SlackAccessor
from mirage.cache.index import IndexCacheStore
from mirage.core.slack import files as slack_files
from mirage.core.slack.history import get_history_jsonl
from mirage.core.slack.users import get_user_profile
from mirage.types import PathSpec
from mirage.utils.errors import enoent


async def read(
    accessor: SlackAccessor,
    path: PathSpec,
    index: IndexCacheStore = None,
) -> bytes:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    virtual = path.original
    prefix = path.prefix if isinstance(path, PathSpec) else ""
    raw = path.original if isinstance(path, PathSpec) else path
    if prefix and raw.startswith(prefix):
        raw = raw[len(prefix):] or "/"
    key = raw.strip("/")
    parts = key.split("/")

    if (len(parts) == 4 and parts[0] in ("channels", "dms")
            and parts[3] == "chat.jsonl"):
        parent_key = f"{parts[0]}/{parts[1]}"
        if index is None:
            raise enoent(virtual)
        virtual_key = prefix + "/" + parent_key
        lookup = await index.get(virtual_key)
        if lookup.entry is None:
            raise enoent(virtual)
        channel_id = lookup.entry.id
        date_str = parts[2]
        return await get_history_jsonl(accessor.config, channel_id, date_str)

    if (len(parts) == 5 and parts[0] in ("channels", "dms")
            and parts[3] == "files"):
        if index is None:
            raise enoent(virtual)
        virtual_key = prefix + "/" + key
        lookup = await index.get(virtual_key)
        if lookup.entry is None or not lookup.entry.extra:
            raise enoent(virtual)
        url = lookup.entry.extra.get("url_private_download")
        if not url:
            raise enoent(virtual)
        return await slack_files.download_file(accessor.config, url)

    if len(parts) == 2 and parts[0] == "users":
        if index is None:
            raise enoent(virtual)
        virtual_key = prefix + "/" + key
        lookup = await index.get(virtual_key)
        if lookup.entry is None:
            raise enoent(virtual)
        user = await get_user_profile(accessor.config, lookup.entry.id)
        return json.dumps(user, ensure_ascii=False,
                          separators=(",", ":")).encode()

    raise enoent(virtual)
